trie/trieStorageManagerInEpoch.go

```diff
@@ -0,0 +1,68 @@
+package trie
+
+import (
+	"fmt"
+
+	"github.com/ElrondNetwork/elrond-go-core/core/check"
+	"github.com/ElrondNetwork/elrond-go/common"
+	"github.com/ElrondNetwork/elrond-go/errors"
+)
+
+type trieStorageManagerInEpoch struct {
+	*trieStorageManager
+	mainStorer dbWithGetFromEpoch
+	epoch      uint32
+}
+
+func newTrieStorageManagerInEpoch(storageManager common.StorageManager, epoch uint32) (*trieStorageManagerInEpoch, error) {
+	if check.IfNil(storageManager) {
+		return nil, ErrNilTrieStorage
+	}
+
+	tsm, ok := storageManager.GetBaseTrieStorageManager().(*trieStorageManager)
+	if !ok {
+		storerType := fmt.Sprintf("%T", storageManager.GetBaseTrieStorageManager())
+		return nil, fmt.Errorf("invalid storage manager, type is %s", storerType)
+	}
+
+	storer, ok := tsm.mainStorer.(dbWithGetFromEpoch)
+	if !ok {
+		storerType := fmt.Sprintf("%T", tsm.mainStorer)
+		return nil, fmt.Errorf("invalid storer, type is %s", storerType)
+	}
+
+	return &trieStorageManagerInEpoch{
+		trieStorageManager: tsm,
+		mainStorer:         storer,
+		epoch:              epoch,
+	}, nil
+}
+
+// Get checks all the storers for the given key, and returns it if it is found
+func (tsmie *trieStorageManagerInEpoch) Get(key []byte) ([]byte, error) {
+	tsmie.storageOperationMutex.Lock()
+	defer tsmie.storageOperationMutex.Unlock()
+
+	if tsmie.closed {
+		log.Debug("trieStorageManagerInEpoch get context closing", "key", key)
+		return nil, errors.ErrContextClosing
+	}
+
+	numEpochsToVerify := uint32(2)
+	for i := uint32(0); i < numEpochsToVerify; i++ {
+		if i > tsmie.epoch {
+			break
+		}
+		epoch := tsmie.epoch - i
+
+		val, err := tsmie.mainStorer.GetFromEpoch(key, epoch)
+		if errors.IsClosingError(err) {
```

Comment:
I think this is not the only case on which we should return the same error. What happens if, for example, we can not get from epoch x because the files from epoch x are in use. Then it will return the error KeyNotFound which is a little bit misleading. I would put the filtering here to continue checking the val's length if the error does not contain "key not found" or something similar (should define that error in the `errors` package and use it in both level db & pruning storer implementations). What do you think?

Reply:
Refactored as discussed.